File: bin/test_proxy.js
#!/usr/bin/env node

/**
 * JavaScript test script for the SOCKS5 proxy server functionality.
 * Tests the proxy using axios with axios-socks5-agent.
 * 
 * Usage: node test_proxy.js
 */

const axios = require('axios');
const { SocksProxyAgent } = require('socks-proxy-agent');

// Proxy configuration
const PROXY_HOST = '127.0.0.1';  // Changed from 0.0.0.0 to 127.0.0.1
const PROXY_PORT = 1080;  // Updated to test new server
const PROXY_USERNAME = 'admin';
const PROXY_PASSWORD = 'TwlNcsmk5MyaVXva';

// Create SOCKS5 proxy URL
const proxyUrl = `socks5://${PROXY_USERNAME}:${PROXY_PASSWORD}@${PROXY_HOST}:${PROXY_PORT}`;

console.log('SOCKS5 Proxy Test (JavaScript)');
console.log('=====================================');
console.log(`Testing proxy: ${proxyUrl}`);

async function testProxyConnection() {
    try {
        // Create SOCKS5 proxy agent using socks-proxy-agent
        const agent = new SocksProxyAgent(proxyUrl);
        
        // Configure axios with the SOCKS5 agent
        const axiosInstance = axios.create({
            httpsAgent: agent,
            httpAgent: agent,
            timeout: 10000
        });

        console.log('\nTesting HTTP request through proxy...');
        
        // Test HTTP request
        const httpResponse = await axiosInstance.get('http://httpbin.org/ip');
        console.log(`✅ HTTP Response status: ${httpResponse.status}`);
        console.log(`   Response data: ${JSON.stringify(httpResponse.data, null, 2)}`);

        console.log('\nTesting HTTPS request through proxy...');
        
        // Test HTTPS request
        const httpsResponse = await axiosInstance.get('https://httpbin.org/ip');
        console.log(`✅ HTTPS Response status: ${httpsResponse.status}`);
        console.log(`   Response data: ${JSON.stringify(httpsResponse.data, null, 2)}`);

        console.log('\n✅ All proxy tests successful!');
        console.log('🎉 SOCKS5 proxy is working correctly with JavaScript/Node.js');
        
        return true;
        
    } catch (error) {
        console.error('❌ Proxy test failed:');
        console.error(`   Error: ${error.message}`);
        
        if (error.code) {
            console.error(`   Error code: ${error.code}`);
        }
        
        if (error.response) {
            console.error(`   Response status: ${error.response.status}`);
            console.error(`   Response data: ${JSON.stringify(error.response.data, null, 2)}`);
        }
        
        return false;
    }
}

async function testWithoutProxy() {
    try {
        console.log('\n📍 Testing direct connection (without proxy)...');
        
        const response = await axios.get('http://httpbin.org/ip', { timeout: 5000 });
        console.log(`   Direct IP: ${response.data.origin}`);
        
    } catch (error) {
        console.error(`   Direct connection failed: ${error.message}`);
    }
}

async function testProxyAuthentication() {
    console.log('\n🔐 Testing proxy authentication...');
    
    try {
        // Test with correct credentials
        console.log('   Testing with correct credentials...');
        const correctAgent = new SocksProxyAgent(proxyUrl);
        const correctAxios = axios.create({
            httpAgent: correctAgent,
            httpsAgent: correctAgent,
            timeout: 5000
        });
        
        await correctAxios.get('http://httpbin.org/ip');
        console.log('   ✅ Authentication with correct credentials: SUCCESS');
        
    } catch (error) {
        console.error('   ❌ Authentication with correct credentials: FAILED');
        console.error(`      Error: ${error.message}`);
    }
    
    try {
        // Test with incorrect credentials
        console.log('   Testing with incorrect credentials...');
        const wrongProxyUrl = `socks5://wrong:credentials@${PROXY_HOST}:${PROXY_PORT}`;
        const wrongAgent = new SocksProxyAgent(wrongProxyUrl);
        const wrongAxios = axios.create({
            httpAgent: wrongAgent,
            httpsAgent: wrongAgent,
            timeout: 5000
        });
        
        await wrongAxios.get('http://httpbin.org/ip');
        console.error('   ❌ Authentication should have failed with wrong credentials');
        
    } catch (error) {
        console.log('   ✅ Authentication correctly rejected wrong credentials');
        console.log(`      Error: ${error.message}`);
    }
}

async function main() {
    console.log(`Node.js version: ${process.version}`);
    console.log('axios: using CommonJS require');
    console.log('socks-proxy-agent: using CommonJS require');
    
    // Run all tests
    await testWithoutProxy();
    
    const proxyWorking = await testProxyConnection();
    
    if (proxyWorking) {
        await testProxyAuthentication();
    }
    
    console.log('\n=====================================');
    if (proxyWorking) {
        console.log('🎯 Overall result: PROXY IS WORKING');
        process.exit(0);
    } else {
        console.log('💥 Overall result: PROXY FAILED');
        process.exit(1);
    }
}

// Handle unhandled promise rejections
process.on('unhandledRejection', (reason, promise) => {
    console.error('Unhandled Rejection at:', promise, 'reason:', reason);
    process.exit(1);
});

// Run the test
main().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
});

File: bin/test_proxy.py
#!/usr/bin/env python3
# coding=utf8

"""
Test script for the SOCKS5 proxy server functionality.
This script can be used to verify the proxy is working correctly.
"""

import requests
import sys
import time
from proxy_server import RinnaProxyServer
import asyncio
import threading
import os


def test_proxy_client(proxy_url: str):
    """Test the proxy by making HTTP requests through it."""
    print(f"Testing proxy: {proxy_url}")
    
    # Configure requests to use the SOCKS5 proxy
    proxies = {
        'http': proxy_url,
        'https': proxy_url
    }
    
    try:
        # Test HTTP request through proxy
        print("Testing HTTP request through proxy...")
        response = requests.get('http://httpbin.org/ip', proxies=proxies, timeout=10)
        print(f"Response status: {response.status_code}")
        print(f"Response content: {response.text}")
        
        # Test HTTPS request through proxy  
        print("\nTesting HTTPS request through proxy...")
        response = requests.get('https://httpbin.org/ip', proxies=proxies, timeout=10)
        print(f"Response status: {response.status_code}")
        print(f"Response content: {response.text}")
        
        print("\n✅ Proxy test successful!")
        return True
        
    except Exception as e:
        print(f"❌ Proxy test failed: {e}")
        return False


def test_proxy_authentication():
    """Test proxy authentication by trying both valid and invalid credentials."""
    host = "127.0.0.1"
    port = 1081  # Use different port for testing
    username = "admin"
    password = "test123"
    
    # Start proxy server in background
    proxy_server = RinnaProxyServer(host=host, port=port, username=username, password=password)
    
    def run_proxy():
        loop = asyncio.new_event_loop()
        asyncio.set_event_loop(loop)
        try:
            loop.run_until_complete(proxy_server.start())
        except Exception as e:
            print(f"Proxy server error: {e}")
        finally:
            loop.close()
    
    proxy_thread = threading.Thread(target=run_proxy, daemon=True)
    proxy_thread.start()
    
    # Wait for server to start
    time.sleep(2)
    
    print("Testing proxy authentication...")
    
    # Test with correct credentials
    proxy_url = f"socks5://{username}:{password}@{host}:{port}"
    print(f"\nTesting with correct credentials: {proxy_url}")
    success = test_proxy_client(proxy_url)
    
    if success:
        print("✅ Authentication test passed")
    else:
        print("❌ Authentication test failed")
    
    # Test with incorrect credentials
    wrong_proxy_url = f"socks5://wrong:credentials@{host}:{port}"
    print(f"\nTesting with incorrect credentials: {wrong_proxy_url}")
    proxies = {'http': wrong_proxy_url, 'https': wrong_proxy_url}
    
    try:
        response = requests.get('http://httpbin.org/ip', proxies=proxies, timeout=5)
        print("❌ Should have failed with wrong credentials")
    except Exception as e:
        print(f"✅ Correctly rejected wrong credentials: {e}")
    
    return success


def main():
    """Main test function."""
    print("SOCKS5 Proxy Server Test")
    print("=" * 40)
    
    if len(sys.argv) > 1:
        if sys.argv[1] == "auth":
            # Test authentication
            test_proxy_authentication()
        elif sys.argv[1] == "client":
            # Test with existing proxy
            proxy_url = sys.argv[2] if len(sys.argv) > 2 else "socks5://admin:rinna_proxy_2025@127.0.0.1:1080"
            test_proxy_client(proxy_url)
        else:
            print("Usage:")
            print("  python test_proxy.py auth           # Test proxy authentication")
            print("  python test_proxy.py client [url]   # Test proxy client with given URL")
    else:
        # Run both tests
        print("Running authentication test...\n")
        test_proxy_authentication()
        
        print("\n" + "=" * 40)
        print("Test completed!")


if __name__ == "__main__":
    # Add parent directory to path to import proxy_server
    sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    main()